core/src/main/java/com/linecorp/armeria/common/RequestContextAwareCompletableFuture.java

```diff
@@ -34,213 +37,405 @@
 
     @Override
     public <U> CompletableFuture<U> thenApply(Function<? super T,? extends U> fn) {
-        return ctx.makeContextAware(super.thenApply(ctx.makeContextAware(fn)));
+        final CompletableFuture<U> result = super.thenApply(ctx.makeContextAware(fn));
+        if (result instanceof RequestContextAwareCompletableFuture) {
```

Comment:
Let's go ahead and extract a private method `maybeMakeContextAware`. We can add a comment like ```If already an instance of Requestcontextawarecompletablefuture, it must either be `this` or the result of `newIncompleteFuture` so we don't have to do anything.```

Reply:
Got it. I missed refactoring point.